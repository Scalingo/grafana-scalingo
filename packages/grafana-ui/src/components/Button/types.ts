import { GrafanaTheme } from '@grafana/data';

export type ButtonVariant = 'primary' | 'secondary' | 'danger' | 'inverse' | 'transparent' | 'destructive' | 'link';

export type ButtonSize = 'xs' | 'sm' | 'md' | 'lg';

export interface StyleDeps {
  theme: GrafanaTheme;
  size: ButtonSize;
  variant: ButtonVariant;
  textAndIcon?: boolean;
}

export interface ButtonStyles {
  button: string;
  iconWrap: string;
  icon?: string;
}
