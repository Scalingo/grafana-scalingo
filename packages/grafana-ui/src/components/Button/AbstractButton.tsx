import React, { AnchorHTMLAttributes, ButtonHTMLAttributes } from 'react';
import tinycolor from 'tinycolor2';
import { css, cx } from 'emotion';
import { Themeable, GrafanaTheme } from '../../types';
import { selectThemeVariant } from '../../themes/selectThemeVariant';

export enum ButtonVariant {
  Primary = 'primary',
  Secondary = 'secondary',
  Danger = 'danger',
  Inverse = 'inverse',
  Transparent = 'transparent',
}

export enum ButtonSize {
  ExtraSmall = 'xs',
  Small = 'sm',
  Medium = 'md',
  Large = 'lg',
  ExtraLarge = 'xl',
}

export interface CommonButtonProps {
  size?: ButtonSize;
  variant?: ButtonVariant;
  /**
   * icon prop is a temporary solution. It accepts lefacy icon class names for the icon to be rendered.
   * TODO: migrate to a component when we are going to migrate icons to @grafana/ui
   */
  icon?: string;
  className?: string;
}

export interface LinkButtonProps extends CommonButtonProps, AnchorHTMLAttributes<HTMLAnchorElement> {}
export interface ButtonProps extends CommonButtonProps, ButtonHTMLAttributes<HTMLButtonElement> {}

interface AbstractButtonProps extends CommonButtonProps, Themeable {
  renderAs: React.ComponentType<CommonButtonProps> | string;
}

const buttonVariantStyles = (
  from: string,
  to: string,
  textColor: string,
  textShadowColor = 'rgba(0, 0, 0, 0.1)',
  invert = false
) => css`
  background: linear-gradient(to bottom, ${from}, ${to});
  color: ${textColor};
  text-shadow: 0 ${invert ? '1px' : '-1px'} ${textShadowColor};
  &:hover {
    background: ${from};
    color: ${textColor};
  }

  &:focus {
    background: ${from};
    outline: none;
  }
`;

const getButtonStyles = (theme: GrafanaTheme, size: ButtonSize, variant: ButtonVariant, withIcon: boolean) => {
  const borderRadius = theme.border.radius.sm;
  let padding,
    background,
    fontSize,
    iconDistance,
    height,
    fontWeight = theme.typography.weight.semibold;

  switch (size) {
    case ButtonSize.Small:
      padding = `${theme.spacing.xs} ${theme.spacing.sm}`;
      fontSize = theme.typography.size.sm;
      iconDistance = theme.spacing.xs;
      height = theme.height.sm;
      break;
    case ButtonSize.Large:
      padding = `${theme.spacing.md} ${theme.spacing.lg}`;
      fontSize = theme.typography.size.lg;
      fontWeight = theme.typography.weight.regular;
      iconDistance = theme.spacing.sm;
      height = theme.height.lg;
      break;
    default:
      padding = `${theme.spacing.sm} ${theme.spacing.md}`;
      iconDistance = theme.spacing.sm;
      fontSize = theme.typography.size.base;
      height = theme.height.md;
  }

  switch (variant) {
    case ButtonVariant.Primary:
      background = buttonVariantStyles(theme.colors.greenBase, theme.colors.greenShade, theme.colors.white);
      break;
    case ButtonVariant.Secondary:
      background = buttonVariantStyles(theme.colors.blueBase, theme.colors.blueShade, theme.colors.white);
      break;
    case ButtonVariant.Danger:
      background = buttonVariantStyles(theme.colors.redBase, theme.colors.redShade, theme.colors.white);
      break;
    case ButtonVariant.Inverse:
      const from = selectThemeVariant({ light: theme.colors.gray5, dark: theme.colors.dark6 }, theme.type) as string;
      const to = selectThemeVariant(
        {
          light: tinycolor(from)
            .darken(5)
            .toString(),
          dark: tinycolor(from)
            .lighten(4)
            .toString(),
        },
        theme.type
      ) as string;

      background = buttonVariantStyles(from, to, theme.colors.link, 'rgba(0, 0, 0, 0.1)', true);
      break;
    case ButtonVariant.Transparent:
      background = css`
        ${buttonVariantStyles('', '', theme.colors.link, 'rgba(0, 0, 0, 0.1)', true)};
        background: transparent;
      `;
      break;
  }

  return {
    button: css`
      label: button;
      display: flex;
      align-items: center;
      font-weight: ${fontWeight};
      font-size: ${fontSize};
      font-family: ${theme.typography.fontFamily.sansSerif};
      line-height: ${theme.typography.lineHeight.xs};
      padding: ${padding};
      text-align: ${withIcon ? 'left' : 'center'};
      vertical-align: middle;
      cursor: pointer;
      border: none;
      height: ${height};
      border-radius: ${borderRadius};
      ${background};

      &[disabled],
      &:disabled {
        cursor: not-allowed;
        opacity: 0.65;
        box-shadow: none;
      }
    `,
    iconWrap: css`
      label: button-icon-wrap;
      display: flex;
      align-items: center;
    `,
    icon: css`
      label: button-icon;
      margin-right: ${iconDistance};
      filter: brightness(100);
    `,
  };
};

export const AbstractButton: React.FunctionComponent<AbstractButtonProps> = ({
  renderAs,
  theme,
  size = ButtonSize.Medium,
  variant = ButtonVariant.Primary,
  className,
  icon,
  children,
  ...otherProps
}) => {
  const buttonStyles = getButtonStyles(theme, size, variant, !!icon);
  const nonHtmlProps = {
    theme,
    size,
    variant,
  };

  const finalClassName = cx(buttonStyles.button, className);
  const finalChildren = icon ? (
    <span className={buttonStyles.iconWrap}>
      <i className={cx([icon, buttonStyles.icon])} />
      <span>{children}</span>
    </span>
  ) : (
    children
  );

  const finalProps =
    typeof renderAs === 'string'
      ? {
          ...otherProps,
          className: finalClassName,
          children: finalChildren,
        }
      : {
          ...otherProps,
          ...nonHtmlProps,
          className: finalClassName,
          children: finalChildren,
        };

  return React.createElement(renderAs, finalProps);
};

AbstractButton.displayName = 'AbstractButton';
