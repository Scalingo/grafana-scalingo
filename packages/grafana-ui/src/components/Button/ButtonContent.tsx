import React from 'react';
import { css } from 'emotion';
import { stylesFactory, useTheme } from '../../themes';
import { GrafanaTheme } from '@grafana/data';
import { ButtonSize } from './types';

const getStyles = stylesFactory((theme: GrafanaTheme) => ({
  content: css`
    display: flex;
    flex-direction: row;
    align-items: center;
    white-space: nowrap;
    height: 100%;
  `,

  icon: css`
    position: relative;
    top: 1px;
    & + * {
      margin-left: ${theme.spacing.sm};
    }
  `,
}));

type Props = {
  icon?: string;
  className?: string;
  children: React.ReactNode;
  size?: ButtonSize;
};

export function ButtonContent(props: Props) {
  const { icon, children } = props;
  const theme = useTheme();
  const styles = getStyles(theme);

  if (!children) {
    return (
      <span className={styles.content}>
        <i className={icon} />
      </span>
    );
  }

  const iconElement = icon && (
    <span className={styles.icon}>
      <i className={icon} />
    </span>
  );

  return (
    <span className={styles.content}>
      {iconElement}
      <span>{children}</span>
    </span>
  );
}
