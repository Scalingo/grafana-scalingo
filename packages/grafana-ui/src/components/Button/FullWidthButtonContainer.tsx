import { css, cx } from '@emotion/css';
import React from 'react';

import { stylesFactory } from '../../themes';

export interface Props {
  className?: string;
}

export const FullWidthButtonContainer = ({ className, children }: React.PropsWithChildren<Props>) => {
  const styles = getStyles();

  return <div className={cx(styles, className)}>{children}</div>;
};

const getStyles = stylesFactory(() => {
  return css`
    display: flex;

    button {
      flex-grow: 1;
      justify-content: center;
    }

    > * {
      flex-grow: 1;
    }

    label {
      flex-grow: 1;
      text-align: center;
    }
  `;
});
