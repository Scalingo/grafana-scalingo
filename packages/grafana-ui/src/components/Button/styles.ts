import tinycolor from 'tinycolor2';
import { css } from 'emotion';
import { selectThemeVariant, stylesFactory } from '../../themes';
import { StyleDeps, ButtonSize } from './types';
import { GrafanaTheme } from '@grafana/data';

const buttonVariantStyles = (
  from: string,
  to: string,
  textColor: string,
  textShadowColor = 'rgba(0, 0, 0, 0.1)',
  invert = false
) => css`
  background: linear-gradient(to bottom, ${from}, ${to});
  color: ${textColor};
  text-shadow: 0 ${invert ? '1px' : '-1px'} ${textShadowColor};
  &:hover {
    background: ${from};
    color: ${textColor};
  }

  &:focus {
    background: ${from};
    outline: none;
  }
`;

export const getButtonStyles = stylesFactory(({ theme, size, variant, textAndIcon }: StyleDeps) => {
  const borderRadius = theme.border.radius.sm;
  const { padding, fontSize, height, fontWeight } = calculateMeasures(theme, size, !!textAndIcon);

  let background;

  switch (variant) {
    case 'primary':
      background = buttonVariantStyles(theme.colors.greenBase, theme.colors.greenShade, theme.colors.white);
      break;

    case 'secondary':
      background = buttonVariantStyles(theme.colors.blueBase, theme.colors.blueShade, theme.colors.white);
      break;

    case 'danger':
      background = buttonVariantStyles(theme.colors.redBase, theme.colors.redShade, theme.colors.white);
      break;

    case 'inverse':
      const from = selectThemeVariant({ light: theme.colors.gray5, dark: theme.colors.dark6 }, theme.type) as string;
      const to = selectThemeVariant(
        {
          light: tinycolor(from)
            .darken(5)
            .toString(),
          dark: tinycolor(from)
            .lighten(4)
            .toString(),
        },
        theme.type
      ) as string;

      background = buttonVariantStyles(from, to, theme.colors.link, 'rgba(0, 0, 0, 0.1)', true);
      break;

    case 'transparent':
      background = css`
        ${buttonVariantStyles('', '', theme.colors.link, 'rgba(0, 0, 0, 0.1)', true)};
        background: transparent;
      `;
      break;

    case 'link':
      background = css`
        ${buttonVariantStyles('', '', theme.colors.linkExternal, 'rgba(0, 0, 0, 0.1)', true)};
        background: transparent;
      `;
      break;
  }

  return {
    button: css`
      label: button;
      display: inline-flex;
      align-items: center;
      font-weight: ${fontWeight};
      font-size: ${fontSize};
      font-family: ${theme.typography.fontFamily.sansSerif};
      line-height: ${theme.typography.lineHeight.xs};
      padding: ${padding};
      vertical-align: middle;
      cursor: pointer;
      border: none;
      height: ${height};
      border-radius: ${borderRadius};
      ${background};

      &[disabled],
      &:disabled {
        cursor: not-allowed;
        opacity: 0.65;
        box-shadow: none;
      }
    `,
    iconWrap: css`
      label: button-icon-wrap;
      & + * {
        margin-left: ${theme.spacing.sm};
      }
    `,
  };
});

type ButtonMeasures = {
  padding: string;
  fontSize: string;
  height: string;
  fontWeight: number;
};

const calculateMeasures = (theme: GrafanaTheme, size: ButtonSize, textAndIcon: boolean): ButtonMeasures => {
  switch (size) {
    case 'sm': {
      return {
        padding: `${theme.spacing.xs} ${theme.spacing.sm}`,
        fontSize: theme.typography.size.sm,
        height: theme.height.sm,
        fontWeight: theme.typography.weight.semibold,
      };
    }

    case 'md': {
      const leftPadding = textAndIcon ? theme.spacing.sm : theme.spacing.md;

      return {
        padding: `${theme.spacing.sm} ${theme.spacing.md} ${theme.spacing.sm} ${leftPadding}`,
        fontSize: theme.typography.size.md,
        height: theme.height.md,
        fontWeight: theme.typography.weight.semibold,
      };
    }

    case 'lg': {
      const leftPadding = textAndIcon ? theme.spacing.md : theme.spacing.lg;

      return {
        padding: `${theme.spacing.md} ${theme.spacing.lg} ${theme.spacing.md} ${leftPadding}`,
        fontSize: theme.typography.size.lg,
        height: theme.height.lg,
        fontWeight: theme.typography.weight.regular,
      };
    }

    default: {
      const leftPadding = textAndIcon ? theme.spacing.sm : theme.spacing.md;

      return {
        padding: `${theme.spacing.sm} ${theme.spacing.md} ${theme.spacing.sm} ${leftPadding}`,
        fontSize: theme.typography.size.base,
        height: theme.height.md,
        fontWeight: theme.typography.weight.regular,
      };
    }
  }
};
