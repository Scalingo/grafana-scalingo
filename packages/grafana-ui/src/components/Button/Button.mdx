import { Meta, Preview, ArgTypes } from '@storybook/blocks';
import { Button, LinkButton } from './Button';
import { Alert } from '../Alert/Alert';

<Meta title="MDX|Button" component={Button} />

# Button

<Alert severity="warning" title={'Please note:'}>
  After reviewing this component we are asking you to use the IconButton when you require a button with only an icon.
</Alert>

When using a button please always follow a11y rules (e.g. W3C Recommendation [3.3.2 Labels or instructions](https://www.w3.org/TR/WCAG21/#labels-or-instructions)) and make sure the context in which the button is located is also communicated by screen readers.

## Primary

Used for "call to action", i.e. triggering the main action. There should never be more than one on a page. If you need multiple buttons for different actions, decide which action is the most important and make that the primary `Button`. All other `Button` components should be secondary.

If there is no primary action, all `Button` components should be secondary.

<Preview>
  <div>
    <Button variant="primary" size="sm" style={{ margin: '5px' }}>
      {'Small'}
    </Button>
    <Button variant="primary" size="md" style={{ margin: '5px' }}>
      {'Medium'}
    </Button>
    <Button variant="primary" size="lg" style={{ margin: '5px' }}>
      {'Large'}
    </Button>
  </div>
</Preview>

## Secondary

The secondary `Button` is the default button style and can trigger various actions. How it is used depends on its surroundings:

1. When next to the primary `Button`, the Secondary style can for example be used for "Cancel" or "Abort" actions.
2. When there is no main important action on a given page, all `Button` components should use the secondary style.

<Preview>
  <div>
    <Button variant="secondary" size="sm" style={{ margin: '5px' }}>
      {'Small'}
    </Button>
    <Button variant="secondary" size="md" style={{ margin: '5px' }}>
      {'Medium'}
    </Button>
    <Button variant="secondary" size="lg" style={{ margin: '5px' }}>
      {'Large'}
    </Button>
  </div>
</Preview>

## Destructive

Used for triggering a removing or deleting action. Because of its dominant coloring, it should be used sparingly. If you need multiple Destructive `Button` components in one view, we recommend using a secondary `Button` or Link variant instead and only use the Destructive variant to double confirm.

<Preview>
  <div>
    <Button variant="destructive" size="sm" style={{ margin: '5px' }}>
      {'Small'}
    </Button>
    <Button variant="destructive" size="md" style={{ margin: '5px' }}>
      {'Medium'}
    </Button>
    <Button variant="destructive" size="lg" style={{ margin: '5px' }}>
      {'Large'}
    </Button>
  </div>
</Preview>

## Text

<Preview>
  <div>
    <Button fill="text" size="sm" style={{ margin: '5px' }}>
      {'Small'}
    </Button>
    <Button fill="text" size="md" style={{ margin: '5px' }}>
      {'Medium'}
    </Button>
    <Button fill="text" size="lg" style={{ margin: '5px' }}>
      {'Large'}
    </Button>
  </div>
</Preview>

<ArgTypes of={Button} />

## Links

To add an anchor that looks like a button use the `<LinkButton>` component and pass a href prop.

<Preview>
  <div>
    <LinkButton href="/" size="sm" style={{ margin: '5px' }}>
      {'Small'}
    </LinkButton>
    <LinkButton href="/" size="md" style={{ margin: '5px' }}>
      {'Medium'}
    </LinkButton>
    <LinkButton href="/" size="lg" style={{ margin: '5px' }}>
      {'Large'}
    </LinkButton>
  </div>
</Preview>
