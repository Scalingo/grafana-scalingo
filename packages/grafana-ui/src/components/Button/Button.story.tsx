import { storiesOf } from '@storybook/react';
import { Button, LinkButton } from './Button';
// @ts-ignore
import withPropsCombinations from 'react-storybook-addon-props-combinations';
import { action } from '@storybook/addon-actions';
import { ThemeableCombinationsRowRenderer } from '../../utils/storybook/CombinationsRowRenderer';
import { boolean } from '@storybook/addon-knobs';
import { getIconKnob } from '../../utils/storybook/knobs';

const ButtonStories = storiesOf('UI/Button', module);

const defaultProps = {
  onClick: [action('Button clicked')],
  children: ['Click click!'],
};

const variants = {
  size: ['xs', 'sm', 'md', 'lg'],
  variant: ['primary', 'secondary', 'danger', 'inverse', 'transparent', 'link'],
};
const combinationOptions = {
  CombinationRenderer: ThemeableCombinationsRowRenderer,
};

const renderButtonStory = (buttonComponent: typeof Button | typeof LinkButton) => {
  const isDisabled = boolean('Disable button', false);
  return withPropsCombinations(
    buttonComponent,
    { ...variants, ...defaultProps, disabled: [isDisabled] },
    combinationOptions
  )();
};

ButtonStories.add('as button element', () => renderButtonStory(Button));

ButtonStories.add('as link element', () => renderButtonStory(LinkButton));

ButtonStories.add('with icon', () => {
  const icon = getIconKnob();
  return withPropsCombinations(
    Button,
    { ...variants, ...defaultProps, icon: [icon && `fa fa-${icon}`] },
    combinationOptions
  )();
});
