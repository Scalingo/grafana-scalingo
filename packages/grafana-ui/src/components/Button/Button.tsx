import React, { AnchorHTMLAttributes, ButtonHTMLAttributes, useContext } from 'react';
import { ThemeContext } from '../../themes';
import { getButtonStyles } from './styles';
import { ButtonContent } from './ButtonContent';
import { ButtonSize, ButtonStyles, ButtonVariant } from './types';
import { cx } from 'emotion';

type CommonProps = {
  size?: ButtonSize;
  variant?: ButtonVariant;
  /**
   * icon prop is a temporary solution. It accepts legacy icon class names for the icon to be rendered.
   * TODO: migrate to a component when we are going to migrate icons to @grafana/ui
   */
  icon?: string;
  className?: string;
  styles?: ButtonStyles;
};

export type ButtonProps = CommonProps & ButtonHTMLAttributes<HTMLButtonElement>;
export const Button = React.forwardRef<HTMLButtonElement, ButtonProps>((props, ref) => {
  const theme = useContext(ThemeContext);
  const { size, variant, icon, children, className, styles: stylesProp, ...buttonProps } = props;

  // Default this to 'button', otherwise html defaults to 'submit' which then submits any form it is in.
  buttonProps.type = buttonProps.type || 'button';

  const styles: ButtonStyles =
    stylesProp ||
    getButtonStyles({
      theme,
      size: size || 'md',
      variant: variant || 'primary',
      textAndIcon: !!(children && icon),
    });

  return (
    <button className={cx(styles.button, className)} {...buttonProps} ref={ref}>
      <ButtonContent icon={icon}>{children}</ButtonContent>
    </button>
  );
});

Button.displayName = 'Button';

export type LinkButtonProps = CommonProps &
  AnchorHTMLAttributes<HTMLAnchorElement> & {
    // We allow disabled here even though it is not standard for a link. We use it as a selector to style it as
    // disabled.
    disabled?: boolean;
  };

export const LinkButton = React.forwardRef<HTMLAnchorElement, LinkButtonProps>((props, ref) => {
  const theme = useContext(ThemeContext);
  const { size, variant, icon, children, className, styles: stylesProp, ...anchorProps } = props;
  const styles: ButtonStyles =
    stylesProp ||
    getButtonStyles({
      theme,
      size: size || 'md',
      variant: variant || 'primary',
      textAndIcon: !!(children && icon),
    });

  return (
    <a className={cx(styles.button, className)} {...anchorProps} ref={ref}>
      <ButtonContent icon={icon}>{children}</ButtonContent>
    </a>
  );
});
LinkButton.displayName = 'LinkButton';
