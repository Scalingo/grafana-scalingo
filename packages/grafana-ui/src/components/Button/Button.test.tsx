import React from 'react';
import { Button, LinkButton } from './Button';
import { mount } from 'enzyme';

describe('Button', () => {
  it('renders correct html', () => {
    const wrapper = mount(<Button icon={'fa fa-plus'}>Click me</Button>);
    expect(wrapper.html()).toMatchSnapshot();
  });
});

describe('LinkButton', () => {
  it('renders correct html', () => {
    const wrapper = mount(<LinkButton icon={'fa fa-plus'}>Click me</LinkButton>);
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('allows a disable state on link button', () => {
    const wrapper = mount(
      <LinkButton disabled icon={'fa fa-plus'}>
        Click me
      </LinkButton>
    );
    expect(wrapper.find('a[disabled]').length).toBe(1);
  });
});
