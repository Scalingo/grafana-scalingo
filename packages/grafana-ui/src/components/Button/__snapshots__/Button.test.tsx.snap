// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders correct html 1`] = `"<button class=\\"css-1kiu3dy-button\\" type=\\"button\\"><span class=\\"css-1beih13\\"><span class=\\"css-1rgbe4\\"><i class=\\"fa fa-plus\\"></i></span><span>Click me</span></span></button>"`;

exports[`LinkButton renders correct html 1`] = `"<a class=\\"css-1kiu3dy-button\\"><span class=\\"css-1beih13\\"><span class=\\"css-1rgbe4\\"><i class=\\"fa fa-plus\\"></i></span><span>Click me</span></span></a>"`;
